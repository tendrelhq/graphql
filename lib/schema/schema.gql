enum AuthenticationProvider {
  legacy
  clerk
}

enum LanguageCode {
  en
  es
}

type Language {
  id: ID!
  code: LanguageCode!
  name: Name!
  name_id: ID!
}

type Name {
  id: ID!
  language: Language!
  language_id: ID!
  value: String!
}

type Tag {
  id: ID!
  name: Name!
  name_id: ID!
  type: String!
  parent: Tag
  parent_id: ID
}

type Customer {
  id: ID!
  active: Boolean!
  name: Name!
  name_id: ID!
  defaultLanguage: Language!
  default_language_id: ID!
}

type Location {
  id: ID!
  active: Boolean!
  activated_at: String
  deactivated_at: String
  name: Name!
  name_id: ID!
  parent: Location
  parent_id: ID
  scan_code: ID
  site: Location!
  site_id: ID!
  tags: [Tag!]!

  children(options: LocationsQueryOptions): [Location!]!
}

type User {
  id: ID!
  active: Boolean!
  activated_at: String
  deactivated_at: String
  authentication_provider: Tag
  authentication_provider_id: ID
  language: Language!
  language_id: ID!
  name: String!
  tags: [Tag!]!
}

type Worker {
  id: ID!
  active: Boolean!
  activated_at: String
  deactivated_at: String
  language: Language!
  language_id: ID!
  role: Tag!
  role_id: ID!
  scan_code: String
  tags: [Tag!]!
  user: User!
  user_id: ID!
}

type Query {
  customers: [Customer!]!
  customer(id: ID!): Customer!
  languages: [Language!]!
  locations(customerId: ID!, options: LocationsQueryOptions): [Location!]!
  location(id: ID!): Location!
  roles: [Tag!]!
  workers(customerId: ID!, options: WorkersQueryOptions): [Worker!]!
}

input LocationsQueryOptions {
  cornerstone: Boolean
  site: Boolean
}

input WorkersQueryOptions {
  active: Boolean
  site: ID
}

type Mutation {
  # Location CRUD
  createLocation(input: CreateLocationInput!): Location!
  updateLocation(input: UpdateLocationInput!): Location!

  # User
  ## CRUD
  createUser(input: CreateUserInput!): User!
  updateUser(input: UpdateUserInput!): User!
  ## Helpers
  acceptInvite(input: AcceptInviteInput!): User!

  # Worker CRUD
  createWorker(input: CreateWorkerInput!): Worker!
  updateWorker(input: UpdateWorkerInput!): Worker!
}

input CreateLocationInput {
  id: ID
  active: Boolean!
  name: CreateNameInput!
  org_id: ID!
  scan_code: ID
}

input UpdateLocationInput {
  id: ID!
  activated_at: String
  deactivated_at: String
  name: UpdateNameInput
  scan_code: ID
}

input CreateUserInput {
  id: ID
  active: Boolean!
  authentication_provider_id: ID
  authentication_identity: ID
  language_id: ID!
  name: String!
}

input UpdateUserInput {
  id: ID!
  authentication_provider_id: ID
  authentication_identity: ID
  language_id: ID!
  name: String!
}

input CreateWorkerInput {
  id: ID
  active: Boolean!
  language_id: ID!
  org_id: ID!
  role_id: ID!
  scan_code: ID
  user_id: ID!
}

input UpdateWorkerInput {
  id: ID!
  activated_at: String
  deactivated_at: String
  language_id: ID!
  role_id: ID!
  scan_code: ID
}

input CreateNameInput {
  id: ID
  active: Boolean!
  language_id: ID!
  value: String!
}

input UpdateNameInput {
  id: ID!
  activated_at: String
  deactivated_at: String
  language_id: ID!
  value: String!
}

input AcceptInviteInput {
  id: ID!
  authentication_identity: ID!
}