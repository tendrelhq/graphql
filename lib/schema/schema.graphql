enum Language {
  en
  es
}

type Tag {
  id: ID!
  name: String!
  type: String!
  parent: ID
}

type Customer {
  id: ID!
  name: String!
  defaultLanguage: Language!
}

type Location {
  id: ID!
  customerId: ID!
  name: String!
  parentId: ID
  tags: [Tag!]!
}

type Query {
  customers: [Customer!]!
  customer(id: ID!): Customer!
  locations(customerId: ID!, parentId: ID): [Location!]!
  sites(customerId: ID!): [Location!]!
}

type Mutation {
  createCustomer(input: CreateCustomerInput!): ID!
  updateCustomer(input: UpdateCustomerInput!): ID!
  deleteCustomer(id: ID!): ID!
  createLocation(input: CreateLocationInput!): ID!
  updateLocation(input: UpdateLocationInput!): ID!
  deleteLocation(id: ID!): ID!
}

input CreateTagInput {
  id: ID!
  name: String!
  type: String!
  parent: ID
}

input UpdateTagInput {
  id: ID!
  name: String
}

input CreateCustomerInput {
  id: ID!
  name: String!
  defaultLanguage: Language!
}

input UpdateCustomerInput {
  id: ID!
  name: String
  defaultLanguage: Language
}

input CreateLocationInput {
  id: ID!
  customerId: ID!
  name: String!
  tags: [CreateTagInput!]!
}

input UpdateLocationInput {
  id: ID!
  name: String
  tags: [UpdateTagInput!]
}
