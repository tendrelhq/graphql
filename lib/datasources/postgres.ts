import type { Request } from "express";
import z from "myzod";
import postgres, { type Fragment } from "postgres";
import { makeActiveLoader } from "./activatable";
import makeAttachmentLoader from "./attachment";
import { makeAuditableLoader } from "./auditable";
import makeCustomerRequestedLanguageLoader from "./crl";
import { makeDescriptionLoader } from "./description";
import { makeDynamicStringLoader } from "./dynamicString";
import makeInvitationLoader from "./invitation";
import makeLanguageLoader from "./language";
import makeLocationLoader from "./location";
import {
  makeDisplayNameLoader,
  makeNameLoader,
  makeNameMetadataLoader,
} from "./name";
import makeOrganizationLoader from "./organization";
import { makeRequirementLoader } from "./requirement";
import { makeSopLoader } from "./sop";
import { makeStatusLoader } from "./status";
import makeTagLoader from "./tag";
import makeUserLoader from "./user";
import makeWorkerLoader from "./worker";

const { DB_MAX_CONNECTIONS, DB_STATEMENT_TIMEOUT } = z
  .object({
    DB_MAX_CONNECTIONS: z.number({ coerce: true }).default(3),
    DB_STATEMENT_TIMEOUT: z.number({ coerce: true }).default(10),
  })
  .parse(process.env, { allowUnknown: true });

export const sql = postgres({
  max: DB_MAX_CONNECTIONS,
  // TODO: this is probably a good idea...
  // connection: {
  //   statement_timeout: DB_STATEMENT_TIMEOUT * 1000, // milliseconds
  // },
});

export type SQL = typeof sql;
export type TxSQL = Parameters<Parameters<typeof sql.begin>[1]>[0];

/**
 * Like Array.prototype.join but for sql.Fragments.
 */
export function join(xs: readonly Fragment[], d: Fragment) {
  return xs.reduce((acc, x, i) => sql`${acc} ${i ? sql`${d} ${x}` : x}`, sql``);
}

/**
 * Type predicate for whether to include a user input in a dynamic update
 * clause. Note that we only check for `undefined`, and not `null`, because we
 * say that `null` indicates we want to set the underlying database column to
 * NULL.
 */
export function shouldUpdate<T>(input?: T, existing?: T): input is T {
  return typeof input !== "undefined" && input !== existing;
}

export function unionAll(xs: readonly Fragment[]) {
  return join(xs, sql`UNION ALL`);
}

export function orm(req: Request) {
  return {
    active: makeActiveLoader(req),
    attachment: makeAttachmentLoader(req),
    auditable: makeAuditableLoader(req),
    crl: makeCustomerRequestedLanguageLoader(req),
    description: makeDescriptionLoader(req),
    displayName: makeDisplayNameLoader(req),
    dynamicString: makeDynamicStringLoader(req),
    invitation: makeInvitationLoader(req),
    language: makeLanguageLoader(req),
    location: makeLocationLoader(req),
    name: makeNameLoader(req),
    nameMetadata: makeNameMetadataLoader(req),
    organization: makeOrganizationLoader(req),
    requirement: makeRequirementLoader(req),
    sop: makeSopLoader(req),
    status: makeStatusLoader(req),
    tag: makeTagLoader(req),
    user: makeUserLoader(req),
    worker: makeWorkerLoader(req),
  };
}

export type ORM = ReturnType<typeof orm>;
