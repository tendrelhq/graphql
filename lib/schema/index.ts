export * from "./__generated__/resolvers.generated";
export * from "./__generated__/typeDefs.generated";

export type Context = {
  authScope?: string;
  languageTypeId: number;
};
