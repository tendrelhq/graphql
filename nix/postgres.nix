{inputs, ...}: {
  imports = [
    inputs.process-compose.flakeModule
  ];
  perSystem = {
    config,
    self',
    pkgs,
    lib,
    ...
  }: {
    devShells.postgres = pkgs.mkShellNoCC {
      name = "tendrelhq/graphql+postgres";
      buildInputs = [
        config.packages.postgresql
        pkgs.perlPackages.TAPParserSourceHandlerpgTAP # pg_prove
        pkgs.python3.pkgs.sqlfmt
        pkgs.sqitchPg
        pkgs.squawk
      ];
    };

    packages = {
      inherit (pkgs) postgresql_14; # beta is on 14.12
      postgresql = pkgs.postgresql_16;
      pgddl = pkgs.stdenv.mkDerivation rec {
        name = "pgddl";
        version = "0.29";
        src = pkgs.fetchFromGitHub {
          owner = "lacanoid";
          repo = name;
          rev = version;
          hash = "sha256-W3G6TGtkj+zXXdGZZR0bmZhsLuFJvuGTlDoo8kL8sf0=";
        };
        buildInputs = with pkgs; [perl config.packages.postgresql];
        postPatch = ''
          patchShebangs .
        '';
        installPhase = ''
          install -D -t $out/share/postgresql/extension *.sql
          install -D -t $out/share/postgresql/extension *.control
        '';
      };
    };

    process-compose.devenv = {
      imports = [
        inputs.services.processComposeModules.default
      ];

      cli.options.no-server = false;

      services.postgres.pg1 = {
        enable = true;
        package = config.packages.postgresql;
        #
        extensions = exts:
          with exts;
          with config.packages; [
            pg_cron
            pgddl
            pgtap
            plpgsql_check
          ];
        hbaConf = lib.mkForce [
          # This one is the default docker bridge network.
          {
            type = "host";
            database = "all";
            user = "all";
            address = "172.17.0.0/16";
            method = "trust";
          }
          # On my machine this is the tendrel_default network. YMMV.
          {
            type = "host";
            database = "all";
            user = "all";
            address = "172.18.0.0/16";
            method = "trust";
          }
        ];
        initialDatabases = [
          {
            name = "dev";
          }
        ];
        settings = {
          listen_addresses = "0.0.0.0"; # so docker0 can connect
          log_statement = "all";
          logging_collector = false;
          shared_preload_libraries = "pg_cron,plpgsql_check";
        };
      };

      services.pgadmin.pga1 = {
        enable = true;
        extraConfig = {
          # Run in desktop mode with the master password disabled.
          MASTER_PASSWORD_REQUIRED = false;
          SERVER_MODE = false;
        };
        initialEmail = "";
        initialPassword = "";
      };

      settings.processes = {
        pga1 = {
          depends_on.pg1.condition = "process_healthy";
        };
      };
    };
  };
}
