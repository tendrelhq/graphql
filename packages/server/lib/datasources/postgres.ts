import postgres, { type Fragment } from "postgres";
import z from "zod";

const { DB_MAX_CONNECTIONS, DB_STATEMENT_TIMEOUT_MS } = z
  .object({
    DB_MAX_CONNECTIONS: z.number({ coerce: true }).default(3),
    DB_STATEMENT_TIMEOUT_MS: z.number({ coerce: true }).default(10_000),
  })
  .parse(process.env);

export const sql = postgres({
  max: DB_MAX_CONNECTIONS,
  connection: {
    // We do some recursive queries and are, in general, not particularly
    // optimized so this is just a last resort to avoid bricking the backend.
    statement_timeout: DB_STATEMENT_TIMEOUT_MS,
  },
  types: {
    bigint: postgres.BigInt,
  },
});

export type Sql = typeof sql;
export type TxSql = Parameters<Parameters<typeof sql.begin>[1]>[0];

/**
 * Like Array.prototype.join but for sql.Fragments.
 */
export function join(xs: readonly Fragment[], d: Fragment) {
  return xs.reduce((acc, x, i) => sql`${acc} ${i ? sql`${d} ${x}` : x}`, sql``);
}

/**
 * Type predicate for whether to include a user input in a dynamic update
 * clause. Note that we only check for `undefined`, and not `null`, because we
 * say that `null` indicates we want to set the underlying database column to
 * NULL.
 */
export function shouldUpdate<T>(input?: T, existing?: T): input is T {
  return typeof input !== "undefined" && input !== existing;
}

export function unionAll(xs: readonly Fragment[]) {
  return join(xs, sql`UNION ALL`);
}
