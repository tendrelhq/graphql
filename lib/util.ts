import type { SortOrder } from "@/schema";
import { type GlobalId, decodeGlobalId } from "@/schema/system";
import { GraphQLError } from "graphql";
import type { Fragment } from "postgres";
import { sql } from "./datasources/postgres";

export function isError<T>(e: T | Error): e is Error {
  return e instanceof Error;
}

export function isValue<T>(v: T | Error): v is T {
  return !isError(v);
}

export function nullish<T>(t: T | null | undefined): t is null | undefined {
  return typeof t === "undefined" || t === null;
}

export class AssertionError extends Error {}

/**
 * Raises an assertion error when the condition does not hold, except in
 * production (and when DISABLE_ASSERTIONS is set) in which case we just log a
 * message instead of throwing.
 */
export function assert(condition: boolean, message = "assertion failed") {
  if (!condition) {
    if (
      // Only fire assertions when in dev/test environments,
      // and only when DISABLE_ASSERTIONS is unset.
      process.env.NODE_ENV !== "production" &&
      typeof process.env.DISABLE_ASSERTIONS === "undefined"
    ) {
      throw new AssertionError(message);
    }

    // Else just log a warning. We don't want to wholly lose our assertions.
    console.debug(`invariant violated: ${message}`);
  }
}

export function assertNonNull<T>(
  value: T | null | undefined,
  message = "Cannot return null for semantically non-nullable field.",
): T {
  if (nullish(value)) {
    throw new Error(message);
  }
  return value;
}

export function assertUnderlyingType(
  expected: string | string[],
  received: string,
) {
  const valid =
    typeof expected === "string"
      ? expected === received
      : expected.some(e => e === received);
  if (!valid) {
    throw new Error(
      `Invalid typename; expected: ${expected}, received: ${received}`,
    );
  }
  return received;
}

export type WithKey<T> = T & { _key: string };

export type PaginationArgs = {
  cursor?: GlobalId | null;
  direction: "forward" | "backward";
  limit: number;
};

type RawPaginationArgs = {
  // forward
  first?: number | null;
  after?: string | null;
  // backward
  last?: number | null;
  before?: string | null;
};

export function buildPaginationArgs(
  args: RawPaginationArgs,
  opts: {
    defaultLimit: number;
    maxLimit: number;
  },
): PaginationArgs {
  validatePaginationArgs(args);

  if (args.first) {
    return {
      cursor: map(args.after, decodeGlobalId),
      direction: "forward",
      limit: Math.min(args.first, opts.maxLimit),
    };
  }

  if (args.last) {
    return {
      cursor: map(args.before, decodeGlobalId),
      direction: "backward",
      limit: Math.min(args.last, opts.maxLimit),
    };
  }

  return {
    direction: "forward",
    limit: opts.defaultLimit,
  };
}

export function sortOrder(s: SortOrder): Fragment {
  switch (s) {
    case "asc":
      return sql`ASC`;
    case "desc":
      return sql`DESC`;
  }
}

export function validatePaginationArgs(args: RawPaginationArgs) {
  if ((args.after && args.before) || (args.first && args.last)) {
    throw new GraphQLError(
      "Invalid pagination arguments. To paginate forward, use the `first` and `after` arguments. To paginate backward, use the `last` and `before` arguments.",
      {
        extensions: {
          code: "BAD_REQUEST",
        },
      },
    );
  }
}

export function map<T, R>(
  t: T,
  fn: (t: NonNullable<T>) => R,
): T extends null | undefined ? T : R {
  if (nullish(t)) {
    return t as T extends null | undefined ? T : R;
  }
  return fn(t as NonNullable<T>) as T extends null | undefined ? T : R;
}

type OrElse<R> = () => R;
export function mapOrElse<T, R>(
  t: T,
  fn: (t: NonNullable<T>) => R,
  orElse: R | OrElse<R>,
): R {
  if (nullish(t)) {
    return typeof orElse === "function" ? (orElse as OrElse<R>)() : orElse;
  }
  return fn(t as NonNullable<T>);
}

export function inspect<T>(t: T) {
  console.debug("t =:", t);
  return t;
}

/**
 * Use this function to compare base64 encoded strings for identity.
 */
export function compareBase64(a: string, b: string): boolean {
  return normalizeBase64(a) === normalizeBase64(b);
}

// Postgres inserts newlines after 76 characters. Apparently it's a thing...
const newlinePattern = /\n/g;

/**
 * Use this function to normalize a base64 encoded string, i.e. remove newline
 * characters that are inserted by Postgres (for example) in certain cases.
 *
 * Note that most base64 implementations simply ignore these seemingly erroneous
 * newlines. As such, this function is mainly useful for (printf) debugging.
 */
export function normalizeBase64(s: string) {
  return s.replace(newlinePattern, "");
}
