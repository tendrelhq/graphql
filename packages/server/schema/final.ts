import { resolvers, typeDefs } from "@/schema";
import { makeExecutableSchema, mergeSchemas } from "@graphql-tools/schema";
import type { GraphQLScalarType } from "graphql";
import { type GqlUrl, register } from "./system/scalars/url";
import { getSchema } from "./v1.schema";

const legacySchema = makeExecutableSchema({ typeDefs, resolvers });
const gratsSchema = getSchema();

export const schema = mergeSchemas({ schemas: [legacySchema, gratsSchema] });

// Custom scalars. See https://grats.capt.dev/docs/docblock-tags/scalars/#serialization-and-parsing-of-custom-scalars.
register(schema.getType("URL") as GraphQLScalarType<GqlUrl, string>);
