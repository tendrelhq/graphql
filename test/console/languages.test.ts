import { afterAll, beforeAll, describe, expect, test } from "bun:test";
import { sql } from "@/datasources/postgres";
import { schema } from "@/schema/final";
import {
  type Customer,
  createTestContext,
  execute,
  paginateQuery,
} from "@/test/prelude";
import { assert, mapOrElse } from "@/util";
import { Faker, base, en } from "@faker-js/faker";
import { createCustomer } from "../app/runtime/prelude/canonical";
import {
  AddLanguageTestDocument,
  ListLanguagesTestDocument,
  PaginateLanguagesTestDocument,
} from "./languages.test.generated";

const ctx = await createTestContext();

const seed = mapOrElse(
  process.env.SEED,
  seed => {
    const s = Number.parseInt(seed);
    assert(Number.isFinite(s), "invalid seed");
    return s;
  },
  Date.now(),
);
const faker = new Faker({ locale: [en, base], seed });

describe("[console] languages", () => {
  let CUSTOMER: Customer;

  test("list", async () => {
    const listQuery = await execute(ctx, schema, ListLanguagesTestDocument, {
      customerId: CUSTOMER.id,
    });
    expect(listQuery.errors).toBeFalsy();
    expect(listQuery.data).toMatchSnapshot();
  });

  test("search", async () => {
    const listQuery = await execute(ctx, schema, ListLanguagesTestDocument, {
      customerId: CUSTOMER.id,
      search: {
        primary: true,
      },
    });
    expect(listQuery.errors).toBeFalsy();
    expect(listQuery.data?.node.__typename).toBe("Organization");
    if (listQuery.data?.node.__typename === "Organization") {
      expect(listQuery.data.node.languages.totalCount).toBe(2);
    }
  });

  test("add", async () => {
    const [{ id: languageId }] = await sql`
      select systaguuid as id
      from public.systag
      where systagparentid = 2 and systagtype = 'fr'
    `;

    const addMutation = await execute(ctx, schema, AddLanguageTestDocument, {
      customerId: CUSTOMER.id,
      languageId: languageId,
    });
    expect(addMutation.errors).toBeFalsy();
    expect(addMutation.data).toMatchSnapshot();

    const listQuery = await execute(ctx, schema, ListLanguagesTestDocument, {
      customerId: CUSTOMER.id,
    });
    expect(listQuery.errors).toBeFalsy();
    expect(listQuery.data).toMatchSnapshot();
  });

  test("paginate", async () => {
    let i = 0;
    for await (const page of paginateQuery({
      async execute(cursor) {
        return await execute(ctx, schema, PaginateLanguagesTestDocument, {
          customerId: CUSTOMER.id,
          first: 1,
          after: cursor,
        });
      },
      next(result) {
        if (result.data?.node.__typename !== "Organization") {
          throw "invariant violated";
        }
        return result.data.node?.languages.pageInfo;
      },
    })) {
      expect(page.errors).toBeFalsy();
      i++;
    }
    expect(i).toBe(3); // 3 total, 1 per page
  });

  beforeAll(async () => {
    await sql.begin(async sql => {
      CUSTOMER = await createCustomer({ faker, seed }, ctx, sql);
      console.log("identity", ctx.auth.userId);
    });
  });

  afterAll(async () => {
    // await cleanup(CUSTOMER.id);

    console.log(`
To reproduce this test:

  SEED=${seed} bun test languages.test --bail
    `);
  });
});
