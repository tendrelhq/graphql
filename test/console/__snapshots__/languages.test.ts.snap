// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`[console] languages list 1`] = `
{
  "node": {
    "__typename": "Organization",
    "languages": {
      "__typename": "EnabledLanguageConnection",
      "edges": [
        {
          "__typename": "EnabledLanguageEdge",
          "node": {
            "__typename": "EnabledLanguage",
            "active": {
              "active": true,
            },
            "language": {
              "code": "en",
            },
            "primary": true,
          },
        },
      ],
      "pageInfo": {
        "__typename": "PageInfo",
        "hasNextPage": false,
        "hasPreviousPage": false,
      },
      "totalCount": 1,
    },
  },
}
`;

exports[`[console] languages add 1`] = `
{
  "enableLanguage": {
    "__typename": "EnabledLanguageEdge",
    "node": {
      "__typename": "EnabledLanguage",
      "active": {
        "__typename": "ActivationStatus",
        "active": true,
      },
      "language": {
        "__typename": "Language",
        "code": "es",
      },
      "primary": false,
    },
  },
}
`;
